File: LeetCode/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        fin = []
        
        return self.addNums(l1, l2, carry, fin)
        
        
    def addNums(self, l1, l2, carry, fin):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if not l1 and not l2 and not carry:
            return fin
 
        val = 0
        if l1:
            val += l1.val
        if l2:
            val += l2.val
            
        val += carry
        
        if val >= 10:
            fin.append(val-10)
            self.addNums(l1.next if l1 else None, l2.next if l2 else None, 1, fin)
        else:
            fin.append(val)
            self.addNums(l1.next if l1 else None, l2.next if l2 else None, 0, fin)
        return fin

File: LeetCode/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int sum = 0;
        int carry = 0;
        
        ListNode resultNode(0);
        ListNode* resultHead = &resultNode;
        ListNode* resultPointer = &resultNode;
        
        while(l1 != NULL || l2 != NULL || carry) {
            sum = carry;
            if(l1!=NULL) sum += l1->val;
            if(l2!=NULL) sum += l2->val;
            if(sum >= 10) {
                carry = 1;
                sum = sum % 10;
            }
            else carry = 0;
            resultPointer->next = new ListNode(sum);
            resultPointer = resultPointer->next;
            if(l1 && l1->next) l1 = l1->next;
            else l1 = NULL;
            if(l2 && l2->next) l2 = l2->next;       
            else l2 = NULL;
        }
        return resultHead->next;
    }
};